system_metrics/system_metrics_client_test.go

```diff
@@ -0,0 +1,827 @@
+	close    bool
+}
+
+func (m *mockStreamClient) Send(data *internal.SystemMetricsData) error {
+	if m.sendErr {
+		return errors.New("problem sending data")
+	}
+	m.mu.Lock()
+	m.data = append(m.data, data)
+	m.mu.Unlock()
+	return nil
+}
+
+func (m *mockStreamClient) CloseAndRecv() (*internal.SystemMetricsResponse, error) {
+	if m.closeErr {
+		return nil, errors.New("problem closing data")
+	}
+	m.mu.Lock()
+	m.close = true
+	m.mu.Unlock()
+	return &internal.SystemMetricsResponse{
+		Id: "ID",
+	}, nil
+}
+
+func (m *mockStreamClient) Header() (metadata.MD, error) {
+	return map[string][]string{}, nil
+}
+func (m *mockStreamClient) Trailer() metadata.MD {
+	return map[string][]string{}
+}
+func (m *mockStreamClient) CloseSend() error {
+	return nil
+}
+func (m *mockStreamClient) Context() context.Context {
+	return nil
+}
+func (m *mockStreamClient) SendMsg(i interface{}) error {
+	return nil
+}
+func (m *mockStreamClient) RecvMsg(i interface{}) error {
+	return nil
+}
+
+func (mc *mockClient) StreamSystemMetrics(_ context.Context, opts ...grpc.CallOption) (internal.CedarSystemMetrics_StreamSystemMetricsClient, error) {
+	if mc.streamCreateErr {
+		return nil, errors.New("problem creating stream")
+	}
+	stream := &mockStreamClient{
+		closeErr: mc.streamCloseErr,
+		sendErr:  mc.streamSendErr,
+	}
+	mc.stream = stream
+	return stream, nil
+}
+
+func (mc *mockClient) CloseMetrics(_ context.Context, in *internal.SystemMetricsSeriesEnd, opts ...grpc.CallOption) (*internal.SystemMetricsResponse, error) {
+	if mc.closeErr {
+		return nil, errors.New("close error")
+	}
+	mc.close = in
+	return &internal.SystemMetricsResponse{
+		Id: "ID",
+	}, nil
+}
+
+type mockServer struct {
+	createErr bool
+	addErr    bool
+	closeErr  bool
+	info      bool
+	data      bool
+	close     bool
+}
+
+func (mc *mockServer) CreateSystemMetricRecord(_ context.Context, in *internal.SystemMetrics) (*internal.SystemMetricsResponse, error) {
+	if mc.createErr {
+		return nil, errors.New("create error")
+	}
+	mc.info = true
+	return &internal.SystemMetricsResponse{
+		Id: "ID",
+	}, nil
+}
+
+func (mc *mockServer) AddSystemMetrics(_ context.Context, in *internal.SystemMetricsData) (*internal.SystemMetricsResponse, error) {
+	if mc.addErr {
+		return nil, errors.New("add error")
+	}
+	mc.data = true
+	return &internal.SystemMetricsResponse{
+		Id: "ID",
+	}, nil
+}
+
+func (mc *mockServer) StreamSystemMetrics(internal.CedarSystemMetrics_StreamSystemMetricsServer) error {
+	return nil
+}
+
+func (mc *mockServer) CloseMetrics(_ context.Context, in *internal.SystemMetricsSeriesEnd) (*internal.SystemMetricsResponse, error) {
+	if mc.closeErr {
+		return nil, errors.New("close error")
+	}
+	mc.close = true
+	return &internal.SystemMetricsResponse{
+		Id: "ID",
+	}, nil
+}
+
+func TestNewSystemMetricsClient(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	srv := &mockServer{}
+	require.NoError(t, startRPCService(ctx, srv, 5000))
```

Comment:
Instead of picking random unique port number constants, can we do something like [this](https://github.com/mongodb/jasper/blob/master/testutil/ports.go) in a testutil package or something that spits out unique numbers to use as ports?